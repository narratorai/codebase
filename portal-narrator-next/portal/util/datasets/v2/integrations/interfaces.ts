export interface IIntegrationsConfig {
  [key: string]: {
    displayName: string
    description?: string
    disableParentGroup?: boolean
  }
}

export type BiToolType = 'looker' | 'tableau' | 'metabase' | 'powerbi' | 'datastudio' | 'domo' | 'otherbi'
