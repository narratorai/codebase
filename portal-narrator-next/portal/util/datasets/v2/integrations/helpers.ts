import _ from 'lodash'
import { INTEGRATIONS_CONFIG, BI_POWER_BI, BI_DATA_STUDIO, BI_OTHER } from 'util/datasets/v2/integrations/constants'
import { BiToolType } from 'util/datasets/v2/integrations/interfaces'

export const getIntegrationConfig = (type: string) => {
  if (_.isEmpty(INTEGRATIONS_CONFIG[type])) {
    return {
      displayName: _.startCase(type),
      disableParentGroup: false,
    }
  }

  return INTEGRATIONS_CONFIG[type]
}

export const makeBiToolLabel = (type: BiToolType) => {
  if (type === BI_POWER_BI) {
    return 'Power BI'
  }

  if (type === BI_DATA_STUDIO) {
    return 'Data Studio'
  }

  if (type === BI_OTHER) {
    return 'Other BI'
  }

  return _.startCase(type)
}
