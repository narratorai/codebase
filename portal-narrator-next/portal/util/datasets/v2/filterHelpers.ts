import { isArray, isFinite, toString, compact, find } from 'lodash'
import moment from 'moment'
import { formatTimeStampUtc, isValidIsoString } from 'util/helpers'

export const OPERATOR_IS_NOT_NULL = 'not_is_null'
export const OPERATOR_IS_NULL = 'is_null'
export const OPERATOR_IN = 'is_in'
export const OPERATOR_NOT_IN = 'not_is_in'
export const OPERATOR_CONTAINS_ANY = 'contains_any'
export const OPERATOR_NOT_CONTAINS_ANY = 'not_contains_any'
export const OPERATOR_IS_EMPTY = 'is_empty'
export const OPERATOR_IS_NOT_EMPTY = 'not_is_empty'

export const OPERATOR_TIME_RANGE = 'time_range'
export const OPERATOR_QUICK_TIME_FILTER = 'quick_time_filter'
export const NO_VALUE_OPERATORS = [OPERATOR_IS_NULL, OPERATOR_IS_NOT_NULL, OPERATOR_IS_EMPTY, OPERATOR_IS_NOT_EMPTY]
export const MULTI_SELECT_OPERATORS = [OPERATOR_IN, OPERATOR_NOT_IN, OPERATOR_CONTAINS_ANY, OPERATOR_NOT_CONTAINS_ANY]

// Given an operator value, what's the operator label:
export const getLabelFromOperatorValue = (operatorValue: string, pluralizeLabels = false) => {
  const allOperatorOptions = [
    ...makeStringOperatorOptions(pluralizeLabels),
    ...makeNumberOperatorOptions(pluralizeLabels),
    ...makeTimeOperatorOptions(pluralizeLabels),
  ]

  return find(allOperatorOptions, ['value', operatorValue])?.label
}

// Given a filter value, return string:
export const makeFilterValueString = (
  filterValue: null | string | number | any[] | moment.Moment,
  timezone: string
) => {
  // moment object --> for timestamp fields
  if (isArray(filterValue)) {
    return JSON.stringify(filterValue)
  }

  if (moment.isMoment(filterValue) || isValidIsoString(filterValue)) {
    return formatTimeStampUtc(filterValue, timezone)
  }

  // string or number --> in most cases
  if (filterValue || isFinite(filterValue)) {
    return toString(filterValue)
  }
}

const makeAllOperatorOptions = (pluralizeLabels: boolean) => {
  return [
    { value: 'equal', label: pluralizeLabels ? 'are equal to' : 'is equal to' },
    { value: 'not_equal', label: pluralizeLabels ? 'are not equal to' : 'is not equal to' },
    { value: OPERATOR_IS_NOT_NULL, label: pluralizeLabels ? 'are not Null' : 'is not Null' },
    { value: OPERATOR_IS_NULL, label: pluralizeLabels ? 'are Null' : 'is Null' },
  ]
}

export const makeBooleanOperatorOptions = (pluralizeLabels: boolean) => {
  return [
    { value: 'equal', label: pluralizeLabels ? 'are equal to' : 'is equal to' },
    { value: OPERATOR_IS_NOT_NULL, label: pluralizeLabels ? 'are not Null' : 'is not Null' },
    { value: OPERATOR_IS_NULL, label: pluralizeLabels ? 'are Null' : 'is Null' },
  ]
}

export const makeStringOperatorOptions = (pluralizeLabels: boolean, allowMultiple = true) => {
  // allowMultiple gives us a flag to disable operators that take multiple values ("contains_any", etc...).
  // For example, we don't want to allow for these operators if the user has selected a field or column as the filter value
  return [
    { value: 'contains', label: pluralizeLabels ? 'contain' : 'contains' },
    { value: 'starts_with', label: pluralizeLabels ? 'start with' : 'starts with' },
    { value: 'ends_with', label: pluralizeLabels ? 'end with' : 'ends with' },
    { value: OPERATOR_IN, label: pluralizeLabels ? 'are in' : 'is in', disabled: !allowMultiple },
    { value: OPERATOR_CONTAINS_ANY, label: pluralizeLabels ? 'contain any' : 'contains any', disabled: !allowMultiple },
    {
      value: OPERATOR_NOT_CONTAINS_ANY,
      label: pluralizeLabels ? 'do not contain any' : 'does not contain any',
      disabled: !allowMultiple,
    },
    { value: 'not_contain', label: pluralizeLabels ? 'do not contain' : 'does not contain' },
    { value: 'not_starts_with', label: pluralizeLabels ? 'do not start with' : 'does not start with' },
    { value: 'not_ends_with', label: pluralizeLabels ? 'do not end with' : 'does not end with' },
    { value: OPERATOR_NOT_IN, label: pluralizeLabels ? 'are not in' : 'is not in', disabled: !allowMultiple },
    { value: OPERATOR_IS_EMPTY, label: pluralizeLabels ? 'are empty strings' : 'is empty string' },
    { value: OPERATOR_IS_NOT_EMPTY, label: pluralizeLabels ? 'are not empty strings' : 'is not empty string' },
    ...makeAllOperatorOptions(pluralizeLabels),
  ]
}

export const makeNumberOperatorOptions = (pluralizeLabels: boolean) => {
  return [
    { value: 'greater_than', label: pluralizeLabels ? 'are greater than' : 'is greater than' },
    { value: 'less_than', label: pluralizeLabels ? 'are less than' : 'is less than' },
    {
      value: 'greater_than_equal',
      label: pluralizeLabels ? 'are greater than or equal to' : 'is greater than or equal to',
    },
    { value: 'less_than_equal', label: pluralizeLabels ? 'are less than or equal to' : 'is less than or equal to' },
    ...makeAllOperatorOptions(pluralizeLabels),
  ]
}

export const makeTimeOperatorOptions = (pluralizeLabels: boolean, allowTimeRange?: boolean) => {
  return compact([
    allowTimeRange
      ? { value: OPERATOR_QUICK_TIME_FILTER, label: pluralizeLabels ? 'are within...' : 'is within...' }
      : undefined,
    allowTimeRange
      ? { value: OPERATOR_TIME_RANGE, label: pluralizeLabels ? 'are within range...' : 'is within range...' }
      : undefined,
    { value: 'greater_than', label: pluralizeLabels ? 'are greater than' : 'is greater than' },
    { value: 'less_than', label: pluralizeLabels ? 'are less than' : 'is less than' },
    {
      value: 'greater_than_equal',
      label: pluralizeLabels ? 'are greater than or equal to' : 'is greater than or equal to',
    },
    { value: 'less_than_equal', label: pluralizeLabels ? 'are less than or equal to' : 'is less than or equal to' },
    ...makeAllOperatorOptions(pluralizeLabels),
  ])
}
