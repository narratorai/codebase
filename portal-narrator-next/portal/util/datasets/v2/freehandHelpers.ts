import _ from 'lodash'
import { getGroupFromContext, getGroupColumns } from './helpers'

import { DatasetContext } from 'util/datasets/interfaces'

//
// These two functions transform the field value from column_id
// to Column Name when displayed and then back again
//
export const freehandStringToColumnName = (
  rawString: string,
  machineContext: DatasetContext,
  groupSlug?: string | null
): string => {
  if (rawString) {
    const group = groupSlug ? getGroupFromContext({ context: machineContext, groupSlug }) : null
    const allColumns = group ? getGroupColumns({ group }) : machineContext.columns
    let convertedValue = rawString

    // sort column ids by length to avoid partially matching an id
    const columns = _.sortBy(allColumns, (val) => val.id.length).reverse()
    columns.forEach((column) => {
      // nosemgrep: javascript.lang.security.audit.detect-non-literal-regexp.detect-non-literal-regexp
      convertedValue = convertedValue.replace(new RegExp(column.id, 'g'), column.label)
    })

    return convertedValue
  }
  return rawString
}

export const replaceColumnLabelWithId = ({ text, label, id }: { text: string; label: string; id: string }): string => {
  // while replacing labels, make sure the label isn't followed by anything other than empty space or: ) , . + - * / %
  // as we loop through here, we replace label with id, which could otherwise could get matched by similar ids:

  // if convertedValue started as: "no_warehouse_with_docs"
  // column A: { id: "no_warehouse_no_docs_d6KSwtWqn", label: "no_warehouse_with_docs" }
  // column B: { id: "no_warehouse_no_docs_uacoVAsVm", label: "no_warehouse_no_docs" }
  // 1st iteration of convertedValue becomes: "no_warehouse_no_docs_d6KSwtWqn"
  // 2nd iteration of convertedValue becomes: "no_warehouse_no_docs_uacoVAsVm_d6KSwtWqn"

  // hence checking for a following character, will stop column B's label from matching column A's id
  // but not stop them if the column was in a function -- which would end it in a ) or has comma or math operator after it
  // check for a . b/c you can do label.asc

  // the column label might have regex special characters like () or ? in it. E.g. Discount Code (All Users).
  // these have to be escaped to ensure the regex literally matches them
  // This will mean that the label will always be literally matched, no matter what's in it
  const labelLiteral = label.replace(/[()[\]?!.*+^${}|\\]/g, (input) => `\\${input}`)

  // return the full given text with any found labels replaced with the id
  return text.replace(new RegExp(`${labelLiteral}(?![^\\)\\s\\,\\+\\-\\*\\/\\%\\.\\]])`, 'g'), `${id}`)
}

export const freehandStringToColumnId = (
  rawString: string,
  machineContext: DatasetContext,
  groupSlug?: string | null
): string => {
  if (rawString) {
    const group = groupSlug ? getGroupFromContext({ context: machineContext, groupSlug }) : null
    const allColumns = group ? getGroupColumns({ group }) : machineContext.columns
    let convertedValue = rawString

    // Due to spaces, column names can be matched multiple ways. We sort by length (longest first) to ensure we match properly.
    // e.g. if we have New Recurring Revenue and Recurring Revenue, matching Recurring Revenue first will
    // give us 'New recurring_revenue' instead of 'new_recurring_revenue'
    const columns = _.sortBy(allColumns, (val) => val.label.length).reverse()
    columns.forEach((column) => {
      convertedValue = replaceColumnLabelWithId({ text: convertedValue, label: column.label, id: column.id })
    })

    return convertedValue
  }
  return rawString
}
