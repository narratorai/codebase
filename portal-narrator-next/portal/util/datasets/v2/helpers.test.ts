import { makeColumnSearchSelectOptions } from './helpers'
import { makeFilterValueString } from './filterHelpers'
import { COLUMN_TYPE_STRING } from '../constants'
import { replaceColumnLabelWithId } from './freehandHelpers'
import graphActivitiesRaw from '../../../../test/fixtures/dataset/graphActivities.json'
import queryDefinitionRaw from '../../../../test/fixtures/dataset/queryDefinition.json'
import { IActivity } from 'graph/generated'
import { IDatasetQueryDefinition } from '../interfaces'

const graphActivities = graphActivitiesRaw as IActivity[]
const queryDefinition = queryDefinitionRaw as unknown as IDatasetQueryDefinition

describe('#makeColumnSearchSelectOptions', () => {
  describe('with no groupSlug', () => {
    it('returns grouped parent columns by activity', () => {
      const groupedColumns = makeColumnSearchSelectOptions({
        activities: graphActivities,
        queryDefinition,
      })

      expect(groupedColumns).toMatchSnapshot()
    })

    describe('with columnTypes limitation', () => {
      it('disabled columns properly', () => {
        const groupedColumns = makeColumnSearchSelectOptions({
          activities: graphActivities,
          queryDefinition,
          columnTypes: [COLUMN_TYPE_STRING],
        })

        expect(groupedColumns).toMatchSnapshot()
      })
    })

    describe('with omitColumnIds limitation', () => {
      it('disabled columns properly', () => {
        const groupedColumns = makeColumnSearchSelectOptions({
          activities: graphActivities,
          queryDefinition,
          omitColumnIds: ['limiting_started_session_3aQ4ZrGo_ts'],
        })

        expect(groupedColumns).toMatchSnapshot()
      })
    })
  })

  describe('with groupSlug', () => {
    it('returns group columns by type', () => {
      const groupedColumns = makeColumnSearchSelectOptions({
        activities: graphActivities,
        queryDefinition,
        groupSlug: 'week',
      })

      expect(groupedColumns).toMatchSnapshot()
    })
  })
})

describe('#replaceColumnLabelWithId', () => {
  it('replaces a label with an id', () => {
    expect(replaceColumnLabelWithId({ text: 'replace', label: 'replace', id: '1234' })).toBe('1234')
  })
  it('replaces a label with a space after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace me', label: 'replace', id: '1234' })).toBe('1234 me')
  })
  it('replaces a label with a ) after label in text', () => {
    expect(replaceColumnLabelWithId({ text: '(replace) me', label: 'replace', id: '1234' })).toBe('(1234) me')
  })

  it('replaces a label with a + after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace+ me', label: 'replace', id: '1234' })).toBe('1234+ me')
  })

  it('replaces a label with a + after label in text even if there is no space between labels and math operator', () => {
    expect(replaceColumnLabelWithId({ text: 'replace+me', label: 'replace', id: '1234' })).toBe('1234+me')
  })

  it('replaces a label with a - after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace- me', label: 'replace', id: '1234' })).toBe('1234- me')
  })

  it('replaces a label with a * after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace* me', label: 'replace', id: '1234' })).toBe('1234* me')
  })

  it('replaces a label with a / after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace/ me', label: 'replace', id: '1234' })).toBe('1234/ me')
  })

  it('replaces a label with a % after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace% me', label: 'replace', id: '1234' })).toBe('1234% me')
  })

  it('replaces a label with a . after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace. me', label: 'replace', id: '1234' })).toBe('1234. me')
  })

  it('replaces a label with a , after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace, me', label: 'replace', id: '1234' })).toBe('1234, me')
  })

  it('replaces a label with a ] after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace] me', label: 'replace', id: '1234' })).toBe('1234] me')
  })

  it('replaces a label with a [ before the label and a ] after label in text', () => {
    expect(replaceColumnLabelWithId({ text: '[replace] me', label: 'replace', id: '1234' })).toBe('[1234] me')
  })

  it('replaces all labels that match', () => {
    expect(replaceColumnLabelWithId({ text: 'replace me replace you', label: 'replace', id: '1234' })).toBe(
      '1234 me 1234 you'
    )
  })
  it('replaces all labels that match even if one has a ) or space after it', () => {
    expect(replaceColumnLabelWithId({ text: 'replace me (replace) you', label: 'replace', id: '1234' })).toBe(
      '1234 me (1234) you'
    )
  })

  it('does not replace a label with a ( after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace(me)', label: 'replace', id: '1234' })).toBe('replace(me)')
  })
  it('does not replace a label with an extra alpha character after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replaced', label: 'replace', id: '1234' })).toBe('replaced')
  })
  it('does not replace a label with an extra special character after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace_', label: 'replace', id: '1234' })).toBe('replace_')
  })
  it('does not replace a label with an extra number after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace2', label: 'replace', id: '1234' })).toBe('replace2')
  })

  it('replaces only labels that match and do not have an alpha character after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace me replaced you', label: 'replace', id: '1234' })).toBe(
      '1234 me replaced you'
    )
  })
  it('replaces only labels that match and do not have a number after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace me replace2 you', label: 'replace', id: '1234' })).toBe(
      '1234 me replace2 you'
    )
  })
  it('replaces only labels that match and do not have a special character after label in text', () => {
    expect(replaceColumnLabelWithId({ text: 'replace me replace_ you', label: 'replace', id: '1234' })).toBe(
      '1234 me replace_ you'
    )
  })

  it('replaces labels containing special characters', () => {
    expect(
      replaceColumnLabelWithId({
        text: 'replace (new) [first] * . $ me replace_ you',
        label: 'replace (new) [first] * . $',
        id: '1234',
      })
    ).toBe('1234 me replace_ you')
  })

  it('replaces labels containing dashes', () => {
    expect(
      replaceColumnLabelWithId({
        text: 'my column replace - version me replace_ you',
        label: 'replace - version',
        id: '1234',
      })
    ).toBe('my column 1234 me replace_ you')
  })
})

const timezone = 'America/Los_Angeles'
describe('#makeFilterValueString', () => {
  it('returns a string for string input', () => {
    expect(makeFilterValueString('testing', timezone)).toBe('testing')
  })

  it('returns a string for 0 input', () => {
    expect(makeFilterValueString(0, timezone)).toBe('0')
  })

  it('returns a string for number input', () => {
    expect(makeFilterValueString(3, timezone)).toBe('3')
  })

  it('returns a formatted string for timestamp', () => {
    expect(makeFilterValueString('2022-06-14T07:00:00.000Z', timezone)).toBe('Jun 14th 2022, 12:00am PDT')
  })
})
