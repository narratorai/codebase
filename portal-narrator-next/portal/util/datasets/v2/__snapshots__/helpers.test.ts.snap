// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#makeColumnSearchSelectOptions with groupSlug returns group columns by type 1`] = `
[
  {
    "disabled": false,
    "label": "week_of_started_session_at",
    "optGroupBy": "Group By Columns",
    "type": "timestamp",
    "value": "week_started_session_at_vRGJ_lHs",
  },
  {
    "disabled": false,
    "label": "total_sessions",
    "optGroupBy": "Metric Columns",
    "type": "number",
    "value": "metrics_total_events",
  },
  {
    "disabled": false,
    "label": "total_viewed_course",
    "optGroupBy": "Metric Columns",
    "type": "number",
    "value": "week_metric_total_viewed_course_2ei6FtOF",
  },
  {
    "disabled": false,
    "label": "total_identified_visitor",
    "optGroupBy": "Metric Columns",
    "type": "number",
    "value": "week_metric_total_identified_visitor_hyUcErjq",
  },
  {
    "disabled": false,
    "label": "avg_courses_viewed_per_session",
    "optGroupBy": "Metric Columns",
    "type": "number",
    "value": "metric_append_viewed_course_oo_2_a_7_hn_feature_1_average_slv_jcl_in",
  },
  {
    "disabled": false,
    "label": "total_paid_visits",
    "optGroupBy": "Metric Columns",
    "type": "number",
    "value": "metric_paid_visit_sum_fou_0_x_kn",
  },
  {
    "disabled": false,
    "label": "total_lead_created",
    "optGroupBy": "Metric Columns",
    "type": "number",
    "value": "week_metric_total_lead_created_4RJOm7yo",
  },
  {
    "disabled": false,
    "label": "spend",
    "optGroupBy": "Spend Columns",
    "type": "float",
    "value": "_spend_column_spend",
  },
  {
    "disabled": false,
    "label": "clicks",
    "optGroupBy": "Spend Columns",
    "type": "integer",
    "value": "_spend_column_clicks",
  },
  {
    "disabled": false,
    "label": "impressions",
    "optGroupBy": "Spend Columns",
    "type": "integer",
    "value": "_spend_column_impressions",
  },
  {
    "disabled": false,
    "label": "percent_viewed_course",
    "optGroupBy": "Computed Columns",
    "type": "float",
    "value": "week_computed_conversion_to_viewed_course_2ei6FtOF",
  },
  {
    "disabled": false,
    "label": "percent_identified_visitor",
    "optGroupBy": "Computed Columns",
    "type": "float",
    "value": "week_computed_conversion_to_identified_visitor_hyUcErjq",
  },
  {
    "disabled": false,
    "label": "cost_per_click",
    "optGroupBy": "Computed Columns",
    "type": "float",
    "value": "week_cost_per_click",
  },
  {
    "disabled": false,
    "label": "cost_per_impression",
    "optGroupBy": "Computed Columns",
    "type": "float",
    "value": "week_cost_per_impression",
  },
  {
    "disabled": false,
    "label": "Cost per Session",
    "optGroupBy": "Computed Columns",
    "type": "number",
    "value": "week_cost_per_session",
  },
  {
    "disabled": false,
    "label": "conversion_rate_to_lead_created",
    "optGroupBy": "Computed Columns",
    "type": "float",
    "value": "week_computed_conversion_to_lead_created_4RJOm7yo",
  },
  {
    "disabled": false,
    "label": "Cost per Session Moving Average",
    "optGroupBy": "Computed Columns",
    "type": "number",
    "value": "week_cost_per_session_moving_average",
  },
]
`;

exports[`#makeColumnSearchSelectOptions with no groupSlug returns grouped parent columns by activity 1`] = `
[
  {
    "disabled": false,
    "label": "started_session_at",
    "optGroupBy": "Started Session",
    "type": "timestamp",
    "value": "limiting_started_session_3aQ4ZrGo_ts",
  },
  {
    "disabled": false,
    "label": "customer",
    "optGroupBy": "Started Session",
    "type": "string",
    "value": "limiting_started_session_3aQ4ZrGo_customer",
  },
  {
    "disabled": false,
    "label": "activity",
    "optGroupBy": "Started Session",
    "type": "string",
    "value": "limiting_started_session_3aQ4ZrGo_feature_1",
  },
  {
    "disabled": false,
    "label": "unique_customer_identifier",
    "optGroupBy": "Started Session",
    "type": "string",
    "value": "limiting_started_session_3aQ4ZrGo_join_customer",
  },
  {
    "disabled": false,
    "label": "ad_source",
    "optGroupBy": "Started Session",
    "type": "string",
    "value": "limiting_started_session_3aQ4ZrGo_ad_source",
  },
  {
    "disabled": false,
    "label": "number_of_courses_viewed",
    "optGroupBy": "Aggregate In Between - Viewed Course",
    "type": "integer",
    "value": "append_viewed_course_Oo2A7HN__feature_1",
  },
  {
    "disabled": false,
    "label": "converted_to_viewed_course_at",
    "optGroupBy": "First In Between - Viewed Course",
    "type": "timestamp",
    "value": "conversion_viewed_course_2ei6FtOF_ts",
  },
  {
    "disabled": false,
    "label": "converted_to_viewed_course",
    "optGroupBy": "First In Between - Viewed Course",
    "type": "boolean",
    "value": "converted_to_viewed_course_2ei6FtOF",
  },
  {
    "disabled": false,
    "label": "days_to_viewed_course",
    "optGroupBy": "First In Between - Viewed Course",
    "type": "number",
    "value": "days_to_viewed_course_2ei6FtOF",
  },
  {
    "disabled": false,
    "label": "converted_to_identified_visitor_at",
    "optGroupBy": "First Ever - Identified Visitor",
    "type": "timestamp",
    "value": "conversion_identified_visitor_hyUcErjq_ts",
  },
  {
    "disabled": false,
    "label": "converted_to_identified_visitor",
    "optGroupBy": "First Ever - Identified Visitor",
    "type": "boolean",
    "value": "converted_to_identified_visitor_hyUcErjq",
  },
  {
    "disabled": false,
    "label": "days_to_identified_visitor",
    "optGroupBy": "First Ever - Identified Visitor",
    "type": "number",
    "value": "days_to_identified_visitor_hyUcErjq",
  },
  {
    "disabled": false,
    "label": "converted_to_lead_created_at",
    "optGroupBy": "First In Between - Lead Created",
    "type": "timestamp",
    "value": "conversion_lead_created_4RJOm7yo_ts",
  },
  {
    "disabled": false,
    "label": "converted_to_lead_created",
    "optGroupBy": "First In Between - Lead Created",
    "type": "boolean",
    "value": "converted_to_lead_created_4RJOm7yo",
  },
  {
    "disabled": false,
    "label": "days_to_lead_created",
    "optGroupBy": "First In Between - Lead Created",
    "type": "number",
    "value": "days_to_lead_created_4RJOm7yo",
  },
  {
    "disabled": false,
    "label": "week_of_started_session_at",
    "optGroupBy": "Computed Columns",
    "type": "timestamp",
    "value": "started_session_at_vRGJ_lHs",
  },
  {
    "disabled": false,
    "label": "day_of_started_session_at",
    "optGroupBy": "Computed Columns",
    "type": "timestamp",
    "value": "started_session_at_ZfgCQg7m",
  },
  {
    "disabled": false,
    "label": "paid_visit",
    "optGroupBy": "Computed Columns",
    "type": "integer",
    "value": "paid_visit",
  },
]
`;

exports[`#makeColumnSearchSelectOptions with no groupSlug with columnTypes limitation disabled columns properly 1`] = `
[
  {
    "disabled": true,
    "label": "started_session_at",
    "optGroupBy": "Started Session",
    "type": "timestamp",
    "value": "limiting_started_session_3aQ4ZrGo_ts",
  },
  {
    "disabled": false,
    "label": "customer",
    "optGroupBy": "Started Session",
    "type": "string",
    "value": "limiting_started_session_3aQ4ZrGo_customer",
  },
  {
    "disabled": false,
    "label": "activity",
    "optGroupBy": "Started Session",
    "type": "string",
    "value": "limiting_started_session_3aQ4ZrGo_feature_1",
  },
  {
    "disabled": false,
    "label": "unique_customer_identifier",
    "optGroupBy": "Started Session",
    "type": "string",
    "value": "limiting_started_session_3aQ4ZrGo_join_customer",
  },
  {
    "disabled": false,
    "label": "ad_source",
    "optGroupBy": "Started Session",
    "type": "string",
    "value": "limiting_started_session_3aQ4ZrGo_ad_source",
  },
  {
    "disabled": true,
    "label": "number_of_courses_viewed",
    "optGroupBy": "Aggregate In Between - Viewed Course",
    "type": "integer",
    "value": "append_viewed_course_Oo2A7HN__feature_1",
  },
  {
    "disabled": true,
    "label": "converted_to_viewed_course_at",
    "optGroupBy": "First In Between - Viewed Course",
    "type": "timestamp",
    "value": "conversion_viewed_course_2ei6FtOF_ts",
  },
  {
    "disabled": true,
    "label": "converted_to_viewed_course",
    "optGroupBy": "First In Between - Viewed Course",
    "type": "boolean",
    "value": "converted_to_viewed_course_2ei6FtOF",
  },
  {
    "disabled": true,
    "label": "days_to_viewed_course",
    "optGroupBy": "First In Between - Viewed Course",
    "type": "number",
    "value": "days_to_viewed_course_2ei6FtOF",
  },
  {
    "disabled": true,
    "label": "converted_to_identified_visitor_at",
    "optGroupBy": "First Ever - Identified Visitor",
    "type": "timestamp",
    "value": "conversion_identified_visitor_hyUcErjq_ts",
  },
  {
    "disabled": true,
    "label": "converted_to_identified_visitor",
    "optGroupBy": "First Ever - Identified Visitor",
    "type": "boolean",
    "value": "converted_to_identified_visitor_hyUcErjq",
  },
  {
    "disabled": true,
    "label": "days_to_identified_visitor",
    "optGroupBy": "First Ever - Identified Visitor",
    "type": "number",
    "value": "days_to_identified_visitor_hyUcErjq",
  },
  {
    "disabled": true,
    "label": "converted_to_lead_created_at",
    "optGroupBy": "First In Between - Lead Created",
    "type": "timestamp",
    "value": "conversion_lead_created_4RJOm7yo_ts",
  },
  {
    "disabled": true,
    "label": "converted_to_lead_created",
    "optGroupBy": "First In Between - Lead Created",
    "type": "boolean",
    "value": "converted_to_lead_created_4RJOm7yo",
  },
  {
    "disabled": true,
    "label": "days_to_lead_created",
    "optGroupBy": "First In Between - Lead Created",
    "type": "number",
    "value": "days_to_lead_created_4RJOm7yo",
  },
  {
    "disabled": true,
    "label": "week_of_started_session_at",
    "optGroupBy": "Computed Columns",
    "type": "timestamp",
    "value": "started_session_at_vRGJ_lHs",
  },
  {
    "disabled": true,
    "label": "day_of_started_session_at",
    "optGroupBy": "Computed Columns",
    "type": "timestamp",
    "value": "started_session_at_ZfgCQg7m",
  },
  {
    "disabled": true,
    "label": "paid_visit",
    "optGroupBy": "Computed Columns",
    "type": "integer",
    "value": "paid_visit",
  },
]
`;

exports[`#makeColumnSearchSelectOptions with no groupSlug with omitColumnIds limitation disabled columns properly 1`] = `
[
  {
    "disabled": true,
    "label": "started_session_at",
    "optGroupBy": "Started Session",
    "type": "timestamp",
    "value": "limiting_started_session_3aQ4ZrGo_ts",
  },
  {
    "disabled": false,
    "label": "customer",
    "optGroupBy": "Started Session",
    "type": "string",
    "value": "limiting_started_session_3aQ4ZrGo_customer",
  },
  {
    "disabled": false,
    "label": "activity",
    "optGroupBy": "Started Session",
    "type": "string",
    "value": "limiting_started_session_3aQ4ZrGo_feature_1",
  },
  {
    "disabled": false,
    "label": "unique_customer_identifier",
    "optGroupBy": "Started Session",
    "type": "string",
    "value": "limiting_started_session_3aQ4ZrGo_join_customer",
  },
  {
    "disabled": false,
    "label": "ad_source",
    "optGroupBy": "Started Session",
    "type": "string",
    "value": "limiting_started_session_3aQ4ZrGo_ad_source",
  },
  {
    "disabled": false,
    "label": "number_of_courses_viewed",
    "optGroupBy": "Aggregate In Between - Viewed Course",
    "type": "integer",
    "value": "append_viewed_course_Oo2A7HN__feature_1",
  },
  {
    "disabled": false,
    "label": "converted_to_viewed_course_at",
    "optGroupBy": "First In Between - Viewed Course",
    "type": "timestamp",
    "value": "conversion_viewed_course_2ei6FtOF_ts",
  },
  {
    "disabled": false,
    "label": "converted_to_viewed_course",
    "optGroupBy": "First In Between - Viewed Course",
    "type": "boolean",
    "value": "converted_to_viewed_course_2ei6FtOF",
  },
  {
    "disabled": false,
    "label": "days_to_viewed_course",
    "optGroupBy": "First In Between - Viewed Course",
    "type": "number",
    "value": "days_to_viewed_course_2ei6FtOF",
  },
  {
    "disabled": false,
    "label": "converted_to_identified_visitor_at",
    "optGroupBy": "First Ever - Identified Visitor",
    "type": "timestamp",
    "value": "conversion_identified_visitor_hyUcErjq_ts",
  },
  {
    "disabled": false,
    "label": "converted_to_identified_visitor",
    "optGroupBy": "First Ever - Identified Visitor",
    "type": "boolean",
    "value": "converted_to_identified_visitor_hyUcErjq",
  },
  {
    "disabled": false,
    "label": "days_to_identified_visitor",
    "optGroupBy": "First Ever - Identified Visitor",
    "type": "number",
    "value": "days_to_identified_visitor_hyUcErjq",
  },
  {
    "disabled": false,
    "label": "converted_to_lead_created_at",
    "optGroupBy": "First In Between - Lead Created",
    "type": "timestamp",
    "value": "conversion_lead_created_4RJOm7yo_ts",
  },
  {
    "disabled": false,
    "label": "converted_to_lead_created",
    "optGroupBy": "First In Between - Lead Created",
    "type": "boolean",
    "value": "converted_to_lead_created_4RJOm7yo",
  },
  {
    "disabled": false,
    "label": "days_to_lead_created",
    "optGroupBy": "First In Between - Lead Created",
    "type": "number",
    "value": "days_to_lead_created_4RJOm7yo",
  },
  {
    "disabled": false,
    "label": "week_of_started_session_at",
    "optGroupBy": "Computed Columns",
    "type": "timestamp",
    "value": "started_session_at_vRGJ_lHs",
  },
  {
    "disabled": false,
    "label": "day_of_started_session_at",
    "optGroupBy": "Computed Columns",
    "type": "timestamp",
    "value": "started_session_at_ZfgCQg7m",
  },
  {
    "disabled": false,
    "label": "paid_visit",
    "optGroupBy": "Computed Columns",
    "type": "integer",
    "value": "paid_visit",
  },
]
`;
